File: csharp/components/Strata.Ninject/Interception/Interceptor.cs
//  ##########################################################################
//  # File Name: Interceptor.cs
//  # Copyright: 2018, Sapientia Systems, LLC.
//  ##########################################################################

using Ninject.Extensions.Interception;
using System;

namespace Strata.Ninject.Interception
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// 
    /// </summary>
    ///  
    public abstract
    class Interceptor:
        IInterceptor
    {
        protected
        Interceptor() {}

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        ///
        public virtual void 
        Intercept(IInvocation invocation)
        {
            DoBefore(invocation);

            try
            {
                DoProceed(invocation);
            }
            catch (Exception e)
            {
                OnException(invocation,e);
                throw;
            }
            finally
            {
                DoAfter(invocation);
            }
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///
        protected virtual void 
        DoBefore(IInvocation invocation) {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///
        protected virtual void
        DoProceed(IInvocation invocation)
        {
            invocation.Proceed();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///
        protected virtual void 
        DoAfter(IInvocation invocation) {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        ///
        protected virtual void 
        OnException(IInvocation invocation,Exception e) {}
    }
}

//  ##########################################################################

File: java/Components/Strata.Application.Core/src/strata/application/core/interception/Interceptor.java
//////////////////////////////////////////////////////////////////////////////
// Interceptor.java
//////////////////////////////////////////////////////////////////////////////

package strata.application.core.interception;

import org.aopalliance.intercept.MethodInterceptor;
import org.aopalliance.intercept.MethodInvocation;

import java.util.concurrent.CompletionStage;

public abstract
class Interceptor
    implements MethodInterceptor
{
    protected
    Interceptor() {}

    @Override
    public Object
    invoke(MethodInvocation invocation)
        throws Throwable
    {
        if (isAsynchronous(invocation))
            return
                new AsynchronousInterceptor(
                    this::doBefore,
                    this::doProceed,
                    this::doAfter,
                    this::onException).invoke(invocation);
        else
            return
                new SynchronousInterceptor(
                    this::doBefore,
                    this::doProceed,
                    this::doAfter,
                    this::onException).invoke(invocation);
    }

    protected boolean
    isAsynchronous(MethodInvocation invocation)
    {
        return
            CompletionStage
                .class
                .isAssignableFrom(
                    invocation
                        .getMethod()
                        .getReturnType());
    }

    protected void
    doBefore(MethodInvocation invocation)
        throws Throwable
    {}

    protected Object
    doProceed(MethodInvocation invocation)
        throws Throwable
    {
        return invocation.proceed();
    }

    protected  void
    doAfter(MethodInvocation invocation)
        throws Throwable
    {}

    protected void
    onException(MethodInvocation invocation,Throwable e)
        throws Throwable
    {}


}

//////////////////////////////////////////////////////////////////////////////
